### DISCUSSION TITLE
pnpm docker monorepo

### DISCUSSION THREAD

**Discussion by FredericLatour** [vercel/next.js#38435] [Help]
Hi,
Has anyone been able to deploy a working Nextjs app in a docker image with Pnpm in a monorepo context ?

I want to deploy a nextjs app that depends on a shared package in a monorepo context.
I'm using the following dockerfile.

```dockerfile
FROM node:16-alpine as Builder
RUN apk add --no-cache libc6-compat
RUN npm install -g pnpm@7.4.0


WORKDIR /app
COPY package.json pnpm-lock.yaml pnpm-workspace.yaml tsconfig*.json ./
COPY ./packages/common ./packages/common/
COPY ./packages/bo ./packages/bo/

RUN pnpm install --frozen-lockfile
RUN pnpm --filter common build
RUN pnpm --filter bo build


FROM node:16.14.0-alpine
RUN apk update && apk --no-cache add tzdata && cp /usr/share/zoneinfo/Europe/Paris /etc/localtime
WORKDIR /app

# Uncomment the following line in case you want to disable telemetry during runtime.
# ENV NEXT_TELEMETRY_DISABLED 1
RUN addgroup --system --gid 1001 nodejs
RUN adduser --system --uid 1001 nextjs

# You only need to copy next.config.js if you are NOT using the default configuration
COPY --from=builder //app/packages/bo/next.config.js ./
COPY --from=builder /app/packages/bo/public ./public
COPY --from=builder /app/packages/bo/package.json ./package.json

# Automatically leverage output traces to reduce image size 
# https://nextjs.org/docs/advanced-features/output-file-tracing
COPY --from=builder --chown=nextjs:nodejs /app/packages/bo/node_modules ./node_modules
COPY --from=builder --chown=nextjs:nodejs /app/packages/bo/.next/standalone ./
COPY --from=builder --chown=nextjs:nodejs /app/packages/bo/.next/static ./.next/static

USER nextjs

EXPOSE 3000

ENV PORT 3000
RUN ls -la
CMD ["node", "server.js"]
```

When trying to run the docker container, I get the following error:

```bash
 docker run -p 3000:3000 nextjs-bo-docker
node:internal/modules/cjs/loader:936
  throw err;
  ^

Error: Cannot find module 'next/dist/server/next-server'
Require stack:
- /app/server.js
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)
    at Function.Module._load (node:internal/modules/cjs/loader:778:27)
    at Module.require (node:internal/modules/cjs/loader:1005:19)
    at require (node:internal/modules/cjs/helpers:102:18)
    at Object.<anonymous> (/app/server.js:4:20)
    at Module._compile (node:internal/modules/cjs/loader:1103:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1155:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:77:12) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [ '/app/server.js' ]
}
```
Using `outputFileTracingRoot` experimental property does not change anything.

I tried using version 12.2.1-canary.5 after reading about some possible fix. It does not change anything.
Thanks for any clue or direction.

**Comment by ijjk** [comment]
Hi, can you provide a reproduction? 

> Using outputFileTracingRoot experimental property does not change anything.

This config needs to be set during build to have an affect, was this the case? If so we can investigate closer to make sure this is working properly. We have existing tests for pnpm with standalone mode.

**Comment by FredericLatour** [comment]
Hi,
Thanks for the feedback
After further investigation, it sounds like my problem could be due to 'Next' just being a symlink that points to a folder that does not exist in the docker image.
I'm certainly missing something here.
Would you point me to the pnpm existing tests you are mentioning? 
I've tried to look for it without any success.

**Comment by kicks321** [comment]
@FredericLatour 

[https://nextjs.org/docs/advanced-features/output-file-tracing#caveats](https://nextjs.org/docs/advanced-features/output-file-tracing#caveats)

**While tracing in monorepo setups, the project directory is used for tracing by default. For next build packages/web-app, packages/web-app would be the tracing root and any files outside of that folder will not be included. To include files outside of this folder you can set experimental.outputFileTracingRoot in your next.config.js.**

```js
// packages/web-app/next.config.js
module.exports = {
  experimental: {
    // this includes files from the monorepo base two directories up
    outputFileTracingRoot: path.join(__dirname, '../../'),
  },
}
```

**Comment by clement91190** [comment]
Did anyone got this working? Would be super useful to see a full example of such a Dockerfile for monorepo with pnpm  + nextjs

**Reply by qhantom** [reply]
Hey @clement91190 👋 

I figured it out for our [open source project](https://github.com/Frachtwerk/essencium-frontend).

In our case, we don't even need to setup the `outputFileTracingRoot` option.

Since we also use pnpm and Next.js and need to deploy the stack standalone, these copy commands made the trick. Here is the Dockerfile (not merged or pushed yet, since we need to make further changes on our branch, unrelated to this issue):

```Dockerfile
FROM node:18-alpine

WORKDIR /app

RUN addgroup --system --gid 1001 nodejs
RUN adduser --system --uid 1001 nextjs

RUN mkdir .next

RUN chown nextjs:nodejs .next

COPY --chown=nextjs:nodejs ./.next/standalone ./
COPY --chown=nextjs:nodejs ./public ./packages/app/public
COPY --chown=nextjs:nodejs ./.next/static ./packages/app/.next/static

USER nextjs

EXPOSE 3000

ENV NODE_ENV=production
ENV NEXT_TELEMETRY_DISABLED=1
ENV HOSTNAME="0.0.0.0"

CMD ["node", "./packages/app/server.js"]
```

**Reply by tbouasli** [reply]
Does this actually work for monorepos? You are not building the internal dependecies in the Dockerfile.

**Reply by qhantom** [reply]
@tbouasli We build the app located inside `packages/app` which gets bundled with all its internal and external dependencies. Those land inside the `.next` folder which we copy over.

**Reply by tbouasli** [reply]
@qhantom Oh so that works for you? I might be doing something wrong then. I will take a closer look at your repo to try and figure it out.  Is `dockerize-app` the branch that's on?

**Reply by qhantom** [reply]
@tbouasli Yes that works for us. Exactly, that's the branch. There hasn't been much activity recently since we faced an issue with the Docker build process inside GitHub Actions vs. locally. That needs more investigation.

**Reply by tbouasli** [reply]
@qhantom Hm, that's why I thought your image might not work then. I believe your Dockerfile only works locally because it is getting your local builds and node_modules, and not installing them and building inside the container. 

Usually you need a `.dockerignore` file referencing `node_modules` and `dist` folders to avoid this issue.

**Reply by qhantom** [reply]
@tbouasli Well, the local build as well as the `node_modules` are built inside the pipeline and those artifacts are passed to the docker build job. So I assume they are available in the job. I can remember that it was an issue with the platform the Docker image was built on. I'm on a Apple silicion chip and the pipeline is executed on a x86 architecture.

As soon as I pick up the topic again, I will inform you about the latest results ℹ️

**Comment by Number16BusShelter** [comment]
Just dropped an example of how to build and deploy a Next.js service from a `pnpm` monorepo in multi-stage Docker builds:

🔗 [https://github.com/Number16BusShelter/nextjs-pnpm-workspace-dockerfile/tree/main](https://github.com/Number16BusShelter/nextjs-pnpm-workspace-dockerfile/tree/main)

Hope it helps!
